### PR TITLE
Add API documentation page for transforms

### PR DESCRIPTION
**Pull request by josh146**
**Context:** Currently, transforms do not appear in the documentation

**Description of the Change:**

- Adds a `doc/code/qml_transforms.rst` page, and sets up autosummary: https://pennylane--1186.org.readthedocs.build/en/1186/code/qml_transforms.html

- Upgrades required Sphinx version to 3.5.3

- Replaces m2r with m2r2, since m2r is not longer maintained.

**Benefits:**

- Transforms now appear in the online API documentation.

- With the Sphinx upgrade, the (many) `docutils` deprecation warnings no longer appear.

- Removes a technical hurdle for contributors, who had issues with the unmaintained m2r package.

**Possible Drawbacks:** Some transforms are imported top-level (`qml.adjoint()`, `qml.metric_tensor()`), and some are not `qml.transforms.invisible()`). Autosummary seems to respect this, and allows for specifying the import location. However, a warning is emitted, which is a bit annoying.

**Related GitHub Issues:** n/a

### CODE REVIEW

File: pennylane/transforms/__init__.py [hunk 1]
```diff
@@ -12,10 +12,50 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 """
-This subpackage contains tape and QNode transforms.
+This subpackage contains QNode, quantum function, device, and tape transforms.
```
Comment (josh146) [inline pennylane/transforms/__init__.py:15]: We can expand this section down the line with more details and examples on how to write your own transform

File: doc/conf.py [hunk 2]
```diff
@@ -56,32 +55,12 @@
...
-#     # first notebook cell in generated Jupyter notebooks
-#     'first_notebook_cell': "%matplotlib inline",
-#     # thumbnail size
-#     'thumbnail_size': (400, 400),
-# }
-
-# Remove warnings that occur when generating the the tutorials
-warnings.filterwarnings("ignore", category=UserWarning, message=r"Matplotlib is currently using agg")
-warnings.filterwarnings("ignore", category=FutureWarning, message=r"Passing \(type, 1\) or '1type' as a synonym of type is deprecated.+")
-warnings.filterwarnings("ignore", category=UserWarning, message=r".+?Compilation using quilc will not be available\.")
```
Comment (Thenerdstation) [inline doc/conf.py:84]: Did we get these working?

File: doc/conf.py [hunk 3]
```diff
@@ -25,7 +25,7 @@
 # -- General configuration ------------------------------------------------
 
 # If your documentation needs a minimal Sphinx version, state it here.
-needs_sphinx = '1.6'
+needs_sphinx = '3.3'
```
Comment (Thenerdstation) [inline doc/conf.py:28]: Ooo big upgrade!

File: pennylane/__init__.py [hunk 4]
```diff
@@ -40,9 +40,14 @@
...
-from pennylane.transforms.adjoint import adjoint
-from pennylane.transforms.control import ctrl, ControlledOperation
+from pennylane.transforms import (
+    adjoint,
+    draw,
+    ControlledOperation,
+    ctrl,
+    measurement_grouping,
+    metric_tensor,
+)
```
Comment (Thenerdstation) [inline pennylane/__init__.py:50]: This will move all of these methods to be directly accessible from `qml`.

+1, I think that's a good idea.

Comment (Thenerdstation) [reply pennylane/__init__.py:50]: There maybe some slight naming confusion.
`qml.transfroms.adjoint` now could mean the `adjoint.py` module or the `adjoint` function.

File: doc/conf.py (hunk 2 above)
Comment (josh146) [reply doc/conf.py:84]: These were warning filters for Sphinx-Gallery; we used to build the demos from within the PL documentation. Eventually we spun this out into its own repo, PennyLaneAI/qml. This is a remnant, I was surprised it was still there!

File: doc/conf.py (hunk 3 above)
Comment (josh146) [reply doc/conf.py:28]: 😆 Figured it was about time

File: pennylane/__init__.py (hunk 4 above)
Comment (josh146) [reply pennylane/__init__.py:50]: Oh good point. I think this import/module naming clash is common all over PennyLane.

Although, I think it is context dependent?

```python
from pennylane.transforms import adjoint # imports the function
from pennylane.transforms.adjoint import adjoint # imports the function from the module

qml.transforms.adjoint # the function
```

File: pennylane/transforms/classical_jacobian.py [hunk 5]
```diff
@@ -19,8 +19,55 @@
...
+    \mathbb{R}^m` between the input QNode arguments :math:`w` and the gate arguments :math:`g`, for
+    a given value of the QNode arguments:
+
+    >>> cjac_fn = qml.transforms.classical_jacobian(circuit)
+    >>> weights = np.array([1., 1., 1.], requires_grad=True)
+    >>> cjac = cjac_fn(weights)
+    >>> print(cjac)
+    [[1. 0. 0.]
+     [1. 0. 0.]
+     [0. 0. 2.]]
```
Comment (Thenerdstation) [inline pennylane/transforms/classical_jacobian.py:57]: What happens if `weight` is a higher dimensional tensor?

**Review by Thenerdstation — APPROVED** [review]

File: pennylane/transforms/classical_jacobian.py (hunk 5 above)
Comment (josh146) [reply pennylane/transforms/classical_jacobian.py:57]: Good question, let me check 🤔

Comment (josh146) [reply pennylane/transforms/classical_jacobian.py:57]: Works perfectly 😅 

```python
import pennylane as qml
from pennylane import numpy as np

dev = qml.device("default.qubit", wires=2)

@qml.qnode(dev)
def circuit(weights):
    qml.RX(weights[0, 0], wires=0)
    qml.RY(weights[0, 1], wires=1)
    qml.RZ(weights[1, 0] ** 2, wires=1)
    return qml.expval(qml.PauliZ(0))
```

where

```pycon
>>> cjac_fn = qml.transforms.classical_jacobian(circuit)
>>> weights = np.array([[1., 1.], [1., 1.]], requires_grad=True)
>>> cjac = cjac_fn(weights)
>>> print(cjac)
[[[1. 0.]
  [0. 0.]]

 [[0. 1.]
  [0. 0.]]

 [[0. 0.]
  [2. 0.]]]
```

Comment (Thenerdstation) [reply pennylane/transforms/classical_jacobian.py:57]: noice

Comment (josh146) [reply pennylane/transforms/classical_jacobian.py:57]: i love it when you don't take an edge case into consideration but it just works

**Pull request merged**